### DISCUSSION TITLE
Scripting or Plugin API

### DISCUSSION THREAD

**Discussion by ShadowFungi** [Orama-Interactive/Pixelorama#544] [Ideas]
## Feature Description
A plugin or scripting API using .gd files that could be optionally downloaded and/or loaded into Pixelorama, as a simple feature (such as adding exportable files formats).

## Why?
It would be nice for people who don't have time to make major features and contributions to Pixelorama, but would still like to contribute to programing simple features. It may also be helpful for those who have limited storage and/or don't need as many features as Pixelorama may provide in the future.

## Draw Backs
Adding this may have major ramifications with many parts of Pixelorama's code.

**Comment by OverloadedOrama** [comment]
I was thinking of creating a plugin system where users can load .pck files, which they can create in Godot. This could allow more freedom than just using simple .gd files, but for now I'm not sure how the API would work. I may experiment on this at some point. As always, ideas and contributions are welcome.

**Comment by dphaldes** [comment]
I would like to talk on a few points about the current Extension API : 
Right now the extensions get access to everything in Pixelorama, from UI elements to editing canvas. This is both a positive and negative. It gives plugin authors freedom to write whatever they want without external constraints of what can/cannot be done and thus the application can truly be "extended" in this way. 
Giving so much freedom can also cause issues where plugin authors might edit core UI elements and override application functions changing code behavior in a completely different part of the application.  Since godot doesn't have access specifiers, there isn't a simple way to deny access to certain parts of the application. One way to counter this is exposing extension helper functions. These functions will take of the complicated / risky parts of code and let plugin authors to make changes without delving too much in the application code.
Example Extension Helpers:

Font API: 
I tried to make a simple font changing plugin. This a snippet from it
```gdscript
extends Node

var global:Node

func _enter_tree() -> void:
	global = get_node("/root/Global")
	if global:
		global.control.theme.default_font.font_data = preload("res://src/Extensions/pixel/ark-pixel-12px-latin.ttf")
		global.control.theme.default_font.size = 8
		

func _exit_tree() -> void:
	if global:
		global.control.theme.default_font.font_data = load("res://assets/fonts/Roboto-Regular.ttf")
		global.control.theme.default_font.size = 12
```
I had to change font size when loading and unloading the extension so that UI size isn't messed up. This can easily be converted into a function call like `extensionApi.load_font( <font options> )` and `extensionApi.reset_font()` (latter is important to promote best practices and making sure that unloading extensions doesn't leave residual settings)

Similar extension helpers can be created for the following: 

- ThemeAPI  - this will allow recoloring of UI elements and change button icons. More safe than directly loading a theme.tres file as that might break the UI
- DialogAPI - this will let authors load external dialogs and make sure they handle things like mouse input handling and screen-dimming correctly. This will also take care of dockable dialogs for the plugin. I will expand on this more in a UI proposal soon.
- MenuAPI - This will let plugins register options under the menu bar and assign callbacks.

This structure can easily be expanded by other apis like ExportApi (for plugins like godot scene exports or other engines/applications/formats) or a ToolApi for custom tools.
